app/admin.go

```diff
@@ -168,32 +169,19 @@ func validateLdapFilter(cfg *model.Config, ldap einterfaces.LdapInterface) *mode
 	return ldap.ValidateFilter(*cfg.LdapSettings.UserFilter)
 }
 
-func (a *App) SaveConfig(cfg *model.Config, sendConfigChangeClusterMessage bool) *model.AppError {
-	oldCfg := a.Config()
-	cfg.SetDefaults()
-	a.Desanitize(cfg)
-
-	if err := cfg.IsValid(); err != nil {
-		return err
-	}
-
-	if err := validateLdapFilter(cfg, a.Ldap); err != nil {
+func (a *App) SaveConfig(newCfg *model.Config, sendConfigChangeClusterMessage bool) *model.AppError {
+	if err := validateLdapFilter(newCfg, a.Ldap); err != nil {
 		return err
 	}
 
-	if *a.Config().ClusterSettings.Enable && *a.Config().ClusterSettings.ReadOnlyConfig {
-		return model.NewAppError("saveConfig", "ent.cluster.save_config.error", nil, "", http.StatusForbidden)
+	oldCfg, err := a.Srv.configStore.Set(newCfg)
+	if err == config.ReadOnlyConfigurationError {
+		return model.NewAppError("saveConfig", "ent.cluster.save_config.error", nil, err.Error(), http.StatusForbidden)
+	} else if err != nil {
+		return model.NewAppError("saveConfig", "app.save_config.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 
-	a.DisableConfigWatch()
-
-	a.UpdateConfig(func(update *model.Config) {
-		*update = *cfg
-	})
 	a.PersistConfig()
```

Comment:
As noted in the summary, I preserve the `PersistConfig` interface only because various tests and business logic currently assume it's possible to `UpdateConfig` and then `ReloadConfig` to recover from the backing store if you haven't called `PersistConfig` in between. I don't think this is actually needed, but I didn't want to tear that apart right now.

The `ReloadConfig` appeared to be superfluous -- and the watcher logic is now internalized to the file store.